aopy/precondition.py

```diff
@@ -1,2 +1,225 @@
+    x += a * 0.5 * np.cos(2 * np.pi * freq[1] * t)
+    x += a * 1.5 * np.cos(2 * np.pi * freq[2] * t)
+    x += a * 20 * np.sin(2 * np.pi * 1.2 * np.sqrt(t))  # noise
+    # x += np.random.normal(scale=np.sqrt(noise_power), size=t.shape)  # noise
+
+    return x, t, f0
+
+def bandpass_butterworth_params(cutoff_low, cutoff_high, fs, order = 4, filter_type = 'bandpass'):
+    '''
+    Args:
+        cutoff_low (int): lower cut-off frequency (in Hz)
+        cutoff_high (int): higher cutoff frequency (in Hz)
+        fs (int): sampling rate (in Hz)
+        order (int): Order of the butter worth filter
+        filter_type (str) : Type of filter. Accepts one of the four values - {‘lowpass’, ‘highpass’, ‘bandpass’, ‘bandstop’}
+
+    Returns:
+        bandpass filter parameters
+
+    '''
+    nyq = 0.5 * fs
+    low = cutoff_low / nyq
+    high = cutoff_high / nyq
+
+    b,a = butter( order, [cutoff_low, cutoff_high], btype=filter_type, fs =fs)
+    return b,a
+
+def bandpass_butterworth_filter_data(data, cutoff_low, cutoff_high, fs, order = 4,filter_type = 'bandpass' ):
+    '''
+    Args:
+        data (array): neural data
+        cutoff_low (float): lower cut-off frequency (in Hz)
+        cutoff_high (float): higher cutoff frequency (in Hz)
+        fs (int): sampling rate (in Hz)
+        order: Order of the butterworth filter
+        filter_type (str) : Type of filter. Accepts one of the four values - {‘lowpass’, ‘highpass’, ‘bandpass’, ‘bandstop’}
+
+    Returns:
+        filtered_data: output bandpass filtered data
+
+    '''
+
+    b,a = butter(order, [cutoff_low, cutoff_high], btype=filter_type, fs= fs)
+    # filtered_data = lfilter(b,a,data)
+    filtered_data = filtfilt(b, a, data)
+    return filtered_data
+
+def bandpass_multitaper_filter_data(data, fs, NW = None, BW= None, adaptive = False, jackknife = True, sides = 'default'):
+    '''
+
+    Args:
+        data (ndarray):  time series data where time axis is assumed to be on the last axis (n_channels , n_samples)
+        fs (float): sampling rate of the signal
+        NW (float): Normalized half bandwidth of the data tapers in Hz
+        BW (float): sampling bandwidth of the data tapers in Hz
+        adaptive (bool): Use an adaptive weighting routine to combine the PSD estimates of different tapers.
+        jackknife (bool): Use the jackknife method to make an estimate of the PSD variance at each point.
+        sides (str): This determines which sides of the spectrum to return.
+
+    Returns:
+        f (ndarray) : Frequency points vector
+        psd_est (ndarray): estimated power spectral density (PSD)
+        nu (ndarray): if jackknife = True; estimated variance of the log-psd. If Jackknife = False; degrees of freedom in a chi square model of how the estimated psd is distributed wrt true log - PSD
+    '''
+    f, psd_mt, nu = tsa.multi_taper_psd(data, fs, NW, BW,  adaptive, jackknife, sides )
+    return f, psd_mt, nu
```

Comment:
either this function should be called get_psd_multitaper() or it should have arguments for low and high frequency cutoffs, then select the appropriate frequencies from the psd estimate, and transform back into time domain. i think both would be useful.

Reply:
renaming this function to get_psd_multitaper() to just get the power spectral density after tapers